### ISSUE TITLE
ZkProgram never finishes compiling

### ISSUE THREAD

**Issue by directcuteo** [o1-labs/o1js#1845]
I have a simple ZkProgram that never finishes compiling.

```
import { Field, SelfProof, verify, ZkProgram } from 'o1js';

export const Add = ZkProgram({
	name: 'add-example',
	publicInput: Field,

	methods: {
		init: {
			privateInputs: [],

			async method(state: Field) {
				state.assertEquals(Field(0));
			},
		},
		addNumber: {
			privateInputs: [SelfProof, Field],

			async method(
				newState: Field,
				earlierProof: SelfProof<Field, void>,
				numberToAdd: Field
			) {
				earlierProof.verify();
				newState.assertEquals(earlierProof.publicInput.add(numberToAdd));
			},
		},
		add: {
			privateInputs: [SelfProof, SelfProof],

			async method(
				newState: Field,
				earlierProof1: SelfProof<Field, void>,
				earlierProof2: SelfProof<Field, void>
			) {
				earlierProof1.verify();
				earlierProof2.verify();
				newState.assertEquals(
					earlierProof1.publicInput.add(earlierProof2.publicInput)
				);
			},
		},
	},
});

export async function zkRollup() {
	console.log('compiling...');

	const { verificationKey } = await Add.compile(); 

	console.log('making proof 0'); // <--------------- This is never executed

	const proof0 = await Add.init(Field(0));

	console.log('making proof 1');

	const proof1 = await Add.addNumber(Field(4), proof0, Field(4));

	console.log('making proof 2');

	const proof2 = await Add.add(Field(4), proof1, proof0);

	console.log('verifying proof 2');
	console.log('proof 2 data', proof2.publicInput.toString());

	const ok = await verify(proof2.toJSON(), verificationKey);
	console.log('ok', ok);
}

```

I am running in the **browser** using `o1js 1.8.0`

**Comment by Trivo25** [comment]
Thanks for bringing this up! Did you manage to successfully compile and prove using nodejs or does the app get stuck there too?

**Comment by directcuteo** [comment]
This example is from the o1js docs examples. I assume it works on node.js if it is part of the docs

**Comment by 45930** [comment]
The code works for me on node.  Can you provide more detail about the failure on browser?

**Comment by directcuteo** [comment]
There is no failure. The Add.compile() function never finishes compiling (several hours waited)

**Comment by 45930** [comment]
Ok, can you share details like what browser, what framework you're using, and any console output in your dev console?

Even better, could you share your repo?

**Comment by directcuteo** [comment]
Sure, here is a repository: [https://github.com/directcuteo/zk-program](https://github.com/directcuteo/zk-program)

```
npm install
npm run build
```

open `dist/index.html` in the browser and check the console

**Comment by Shigoto-dev19** [comment]
This issue has been resolved with the merge of #1857 and the fix is now available in o1js npm version `1.9.0`

**Comment by coldstar1993** [comment]
I tried simplifying the code, and find that the issue always happens when there is recursive proof within zkProgram, but smoothly without recursion cases.
